### ISSUE TITLE
shoudlComponentUpdate confusion.

### ISSUE THREAD

**Issue by ogostos** [reactjs/reactjs.org#1579]
According to docs on [shouldComponentUpdate lifecycle](https://github.com/reactjs/reactjs.org/issues/new):

> If you are confident you want to write it by hand, you may compare `this.props` with `nextProps` and `this.state` with `nextState` and return `false` to tell React the update can be skipped. Note that returning `false` **does not prevent child components from re-rendering** when _their state changes_.

Later on, in [optimizing perfomance](https://reactjs.org/docs/optimizing-performance.html#shouldcomponentupdate-in-action) we have:

> Since `shouldComponentUpdate` returned `false` for the subtree rooted at _C2_, React did not attempt to render _C2_, and thus didn’t even have to invoke `shouldComponentUpdate` on _C4_ and _C5_.

Thus, from first part we get the idea that `shouldComponentUpdate` lifecycle opeartes in the domain of **parent component** and has nothing to do with its **children**. Second piece of information states that parents' `shouldComponentUpdate`, which is set to `false` by hand, somehow _**interferes and stops**_ calling that lifecycle further down.

Am I missing something or is there a conflict in docs? Any clarification on this point is very helpful. 
Thanks in advance.

**Comment by sylvix** [comment]
React components are updated when their state or props change. After that render() is called and EVERY component is updated that is returned from that render(). And also components that are returned from those "children" renders too. Let's call it going down recursively the component tree and updating everything. React works this way, it can't divine for itself that something should not be updated.

So, in the docs I think it means that if the change is made from higher positioned component in the tree and shouldComponentUpdate in a component somewhere along the way down returned false, anything lower than that component won't be re-rendered in that update cycle. But, if any of those lower components have their own state and means to change it (ex. buttons, mouse events, etc) - they will update on their own regardless of shouldComponentUpdate higher in the hierarchy.

**Comment by ogostos** [comment]
@sylvix Thanks for your response.
From different perspective, we can have two state in parent's `shouldComponentUpdate`: **`true`** and **`false`**.

In case of **`true`**: re-render _parent_ component including _all of its descendants_*
In case of **`false`**: no need to re-render _parent_ component including _all of its descendants_**

_* if any of descendants have `shouldComponentUpdate` set to **false** - don't re-render them_
_** if any of descendants have `shouldComponentUpdate` set to **true** - re-render them_

As you see, `shouldComponentUpdate` influences only its component. I think this statement:

> Since shouldComponentUpdate returned false for the subtree rooted at C2, React did not attempt to render C2, and thus didn’t even have to invoke shouldComponentUpdate on C4 and C5.

is somewhat misleading and need to be more precise, cause React didn't attempt to render `C2`, but it should have to check `shouldComponentUupdate` on `C4` and `C5`. The logic of automatic exclusion of this check is not following rule discussed above.

**Comment by gaearon** [comment]
@ogostos I don't think your analysis is correct.

React renders from the point you `setState` down, all the way. `shouldComponentUpdate` skips rendering. If a component doesn't render, its children don't render either.

*However* if somewhere inside those skipped trees, children also called `setState`, it also flows from those points all the way down. (Unless, of course, updates get stopped somewhere else.)

In other words, `setState`s are where the flow starts and `shouldComponentUpdate: false` is where it's interrupted. But there can be multiple of either in the same tree.

**Comment by gaearon** [comment]
I think the reason for your confusion is that whether `shouldComponentUpdate` is respected depends on **where the update originated** (`setState`).

The first paragraph you quoted says:

>If you are confident you want to write it by hand, you may compare this.props with nextProps and this.state with nextState and return false to tell React the update can be skipped. Note that returning false does not prevent child components from re-rendering when *their state* changes.

Note italics around *their state*. A parent might say "don't update me" but **if the child calls `setState`, we don't ask the parent. It shouldn't have any control over the internal state of the child and it can't possibly know whether it needs updating or not.

The second paragraph you quoted says:

>Since shouldComponentUpdate returned false for the subtree rooted at C2, React did not attempt to render C2, and thus didn’t even have to invoke shouldComponentUpdate on C4 and C5.

This was about an update **from above**. An update started above, C2 skipped it, and so we didn't even get to asking C4 or C5. This is because C2, being a parent, has enough knowledge to tell whether everything below it needs to update or not in response to some change.

A parent can't decide about updates to child's internal state. But a parent can prevent updates passing *through* it above.
